src/Data/Arena.hs

```diff
@@ -15,74 +17,169 @@ module Data.Arena (
+import           Data.Bits (complement, popCount, (.&.))
 import           Data.Primitive.ByteArray
 import           Data.Primitive.MutVar
+import           Data.Primitive.MVar
+import           Data.Primitive.PrimVar
 
 #ifdef NO_IGNORE_ASSERTS
-import           Control.Monad (forM_)
 import           Data.Word (Word8)
 #endif
 
-data ArenaManager s = ArenaManager
+data ArenaManager s = ArenaManager (MutVar s [Arena s])
 
 newArenaManager :: PrimMonad m => m (ArenaManager (PrimState m))
 newArenaManager = do
-    _toUseTheConstraint <- newMutVar 'x'
-    return ArenaManager
+    m <- newMutVar []
+    return $ ArenaManager m
 
 -- | For use in bencmark environments
 instance NFData (ArenaManager s) where
-    rnf _ = ()
+    rnf (ArenaManager !_) = ()
 
--- TODO: this is debug implementation
--- we retain all the allocated arrays,
--- so we can scramble them at the end.
-data Arena s = Arena (MutVar s [MutableByteArray s])
+data Arena s = Arena
+    { curr :: !(MVar s (Block s))   -- current block, also acts as a lock
+    , free :: !(MutVar s [Block s])
+    , full :: !(MutVar s [Block s])
+    }
+
+data Block s = Block !(PrimVar s Int) !(MutableByteArray s)
 
 instance NFData (Arena s) where
-  rnf (Arena mvar) = rwhnf mvar
+  rnf (Arena !_ !_ !_) = ()
 
 type Size      = Int
 type Offset    = Int
 type Alignment = Int
 
+blockSize :: Int
+blockSize = 0x100000
+
+newBlock :: PrimMonad m => m (Block (PrimState m))
+newBlock = do
+    off <- newPrimVar 0
+    mba <- newAlignedPinnedByteArray blockSize 4096
+    return (Block off mba)
+
 withArena :: PrimMonad m => ArenaManager (PrimState m) -> (Arena (PrimState m) -> m a) -> m a
 withArena am f = do
-  a <- newArena am
-  x <- f a
-  closeArena a
-  pure x
+    a <- newArena am
+    x <- f a
+    closeArena am a
+    pure x
 
 newArena :: PrimMonad m => ArenaManager (PrimState m) -> m (Arena (PrimState m))
-newArena _ = do
-    mvar <- newMutVar []
-    pure $! (Arena mvar)
+newArena (ArenaManager arenas) = do
+    marena <- atomicModifyMutVar' arenas $ \case
+        []     -> ([], Nothing)
+        (x:xs) -> (xs, Just x)
 
-closeArena :: PrimMonad m => Arena (PrimState m) -> m ()
-#ifdef NO_IGNORE_ASSERTS
-closeArena (Arena mvar) = do
-    -- scramble the allocated bytearrays,
-    -- they shouldn't be in use anymore!
-    mbas <- readMutVar mvar
-    forM_ mbas $ \mba -> do
-        size <- getSizeofMutableByteArray mba
-        setByteArray mba 0 size (0x77 :: Word8)
+    case marena of
+        Just arena -> return arena
+        Nothing -> do
+            curr <- newBlock >>= newMVar
+            free <- newMutVar []
+            full <- newMutVar []
+            return Arena {..}
+
+closeArena :: PrimMonad m => ArenaManager (PrimState m) -> Arena (PrimState m) -> m ()
+closeArena (ArenaManager arenas) arena = do
+    scrambleArena arena
+
+    -- reset the arena to clear state
+    resetArena arena
+
+    atomicModifyMutVar' arenas $ \xs -> (arena : xs, ())
```

Comment:
In debug mode, we could even not put back the arena, always handing out fresh arenas. This way, the next user can't overwrite the scrambled blocks again with something that looks like sensible data. On the other hand, this would mean that `newBlockWithFree` doesn't get good test coverage. What do you think?

Reply:
> What do you think?

Exactly, as you

> On the other hand, this would mean that newBlockWithFree doesn't get good test coverage. 

---

I think that if new stuff overwrites something old, some test should fail without scrambling too. So  I chose the approach which is most similar between debug and non-debug mode.

Reply:
Makes sense!